go/store/nbs/archive_writer.go

```diff
@@ -0,0 +1,397 @@
+	}
+
+	if len(b) == 0 {
+		return 0, nil
+	}
+
+	offset := aw.bytesWritten
+
+	written, err := aw.output.Write(b)
+	if err != nil {
+		return 0, err
+	}
+	if written != len(b) {
+		return 0, io.ErrShortWrite
+	}
+	aw.bytesWritten += uint64(written)
+
+	aw.stagedBytes = append(aw.stagedBytes, byteSpan{offset, uint64(written)})
+
+	return uint32(len(aw.stagedBytes)), nil
+}
+
+func (aw *archiveWriter) chunkSeen(h hash.Hash) bool {
+	return aw.seenChunks.Has(h)
+}
+
+func (aw *archiveWriter) stageChunk(hash hash.Hash, dictionary, data uint32) error {
+	if aw.workflowStage != stageByteSpan {
+		return fmt.Errorf("Runtime error: stageChunk called out of order")
+	}
+
+	if data == 0 || data > uint32(len(aw.stagedBytes)) {
+		return ErrInvalidChunkRange
+	}
+	if aw.seenChunks.Has(hash) {
+		return ErrDuplicateChunkWritten
+	}
+	aw.seenChunks.Insert(hash)
+
+	if dictionary > uint32(len(aw.stagedBytes)) {
+		return ErrInvalidDictionaryRange
+	}
+
+	aw.stagedChunks = append(aw.stagedChunks, stagedChunkRef{hash, dictionary, data})
+	return nil
+}
+
+func (scrs stagedChunkRefSlice) Len() int {
+	return len(scrs)
+}
+func (scrs stagedChunkRefSlice) Less(i, j int) bool {
+	return scrs[i].hash.Prefix() < scrs[j].hash.Prefix()
+}
+func (scrs stagedChunkRefSlice) Swap(i, j int) {
+	scrs[i], scrs[j] = scrs[j], scrs[i]
+}
+
+func (aw *archiveWriter) finalizeByteSpans() error {
+	if aw.workflowStage != stageByteSpan {
+		return fmt.Errorf("Runtime error: finalizeByteSpans called out of order")
+	}
+
+	// Get the checksum for the data written so far
+	aw.dataCheckSum = sha512Sum(aw.output.GetSum())
+	aw.output.ResetHasher()
+	aw.workflowStage = stageIndex
+
+	return nil
+}
+
+// writeIndex writes the index to the archive. Expects the hasher to be reset before be called, and will reset it. It
+// sets the indexLen and indexCheckSum fields on the archiveWriter, and updates the bytesWritten field.
+func (aw *archiveWriter) writeIndex() error {
+	if aw.workflowStage != stageIndex {
+		return fmt.Errorf("Runtime error: writeIndex called out of order")
+	}
+
+	startingByteCount := aw.bytesWritten
+
+	varIbuf := make([]byte, binary.MaxVarintLen64)
+
+	// Write out the stagedByteSpans
+	for _, bs := range aw.stagedBytes {
+		n := binary.PutUvarint(varIbuf, bs.length)
+		written, err := aw.output.Write(varIbuf[:n])
+		if err != nil {
+			return err
+		}
+		if written != n {
+			return io.ErrShortWrite
+		}
+		aw.bytesWritten += uint64(written)
+	}
+
+	// sort stagedChunks by hash.Prefix(). Note this isn't a perfect sort for hashes, we are just grouping them by prefix
+	sort.Sort(aw.stagedChunks)
+
+	// We lay down the sorted chunk list in it's three forms.
+	// Prefix Map
+	for _, scr := range aw.stagedChunks {
+		err := aw.writeUint64(scr.hash.Prefix())
+		if err != nil {
+			return err
+		}
+	}
+	// ChunkReferences
+	for _, scr := range aw.stagedChunks {
+		n := binary.PutUvarint(varIbuf, uint64(scr.dictionary))
```

Comment:
Worth making a `aw.writeUVarint64` helper for these three callers? It can avoid heap allocation with a stack array: `var buf [binary.MaxVarintLen64]byte`